app/services/products_renderer.rb

```diff
@@ -0,0 +1,92 @@
+require 'open_food_network/scope_product_to_hub'
+
+class ProductsRenderer
+  class NoProducts < RuntimeError; end
+
+  def initialize(distributor, order_cycle, customer, args = {})
+    @distributor = distributor
+    @order_cycle = order_cycle
+    @customer = customer
+    @args = args
+  end
+
+  def products_json
+    raise NoProducts unless order_cycle && distributor && products
+
+    ActiveModel::ArraySerializer.new(products,
+                                     each_serializer: Api::ProductSerializer,
+                                     current_order_cycle: order_cycle,
+                                     current_distributor: distributor,
+                                     variants: variants_for_shop_by_id,
+                                     master_variants: master_variants_for_shop_by_id,
+                                     enterprise_fee_calculator: enterprise_fee_calculator).to_json
+  end
+
+  private
+
+  attr_reader :order_cycle, :distributor, :customer, :args
+
+  def products
+    return unless order_cycle
+
+    @products ||= begin
+      results = distributed_products.products_relation.order(taxon_order)
+
+      filter_and_paginate(results).
+        each { |product| product_scoper.scope(product) } # Scope results with variant_overrides
+    end
+  end
+
+  def product_scoper
+    OpenFoodNetwork::ScopeProductToHub.new(distributor)
+  end
+
+  def enterprise_fee_calculator
+    OpenFoodNetwork::EnterpriseFeeCalculator.new distributor, order_cycle
+  end
+
+  def filter_and_paginate(query)
+    query.ransack(args[:q]).result.page(args[:page] || 1).per(args[:per_page] || 10)
```

Comment:
We should move these magic numbers to constants but I'd first check whether the pagination gem we use (I don't remember which one of the available) and if it allows defining a global default. Didn't we implement something like that not long ago?

Reply:
Constants added :+1:

I'm actually not sure about messing with global pagination defaults. I think we can start getting unexpected results, like limits being applied "automagically" on queries that will return incorrect results when limited.